rfcs/convergence/contract-for-trigger-pattern.md

```diff
@@ -0,0 +1,105 @@
+# RFC: Creating a structured contract for Trigger - like patterns
```

Comment:
It seems like this RFC is trying to do two things at once:
1. Establish a pattern for trigger elements, and how to specify the props that are passed to them.
2. Sort out the dependency problem when we have a contract between two components like Menu and MenuButton, but don't want to have an explicit dependency.

I agree with (1), and I'm in favor of being explicit about the contract between a component and its trigger, via TriggerProps or otherwise.

For (2), there are some other possible solutions I could think of, which don't require creating a new package for the contract interfaces.
- Have each component define its own contracts separately. And if we want to make sure that two of our components work together, then write a test to make sure of that, rather than having an explicit dependency.
  * Using Tooltip as an example, it defines its contract as requiring that the trigger support a couple of props from HTMLElement, which almost all of our components already do. The components themselves don't need to explicitly have a dependency on TooltipTriggerProps.
  * In the case of Menu and MenuButton: they would each define their own contracts specifying what they require of the component they use: Menu defines what it requires from a trigger, and MenuButton defines what it requires from a menu. This allows Menu to be able to be triggered by any arbitrary button (not only our MenuButton); and for MenuButton to be able to trigger any arbitrary menu (not only our Menu), as long as they support their respective contracts. By being explicit about the contracts, it'll be more straightforward how someone can implement their own version of each.
- Or, if MenuButton is only ever intended to open our Menu, then we should create a new package react-menubutton, and move MenuButton/SplitButton into that package.

Reply:
> It seems like this RFC is trying to do two things at once:

I agree and I think that even the trigger contract could be separated.

For example, like [`useTriggerElement`](https://github.com/microsoft/fluentui/blob/c1f191174845554a35702f43f36a0019dd93065a/packages/fluentui/react-bindings/src/hooks/useTriggerElement.tsx) in Northstar as we should perform numerous checks to ensure that a passed trigger can be at least cloned and props could be applied 🤔